In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
## for feature selection
from sklearn.linear_model import Lasso
from sklearn.feature_selection import SelectFromModel
#to visualise all the columns in the dataframe
pd.pandas.set_option('display.max_columns',None)

In [2]:
dataset=pd.read_csv('train_X.csv')

In [3]:
dataset.head()

Unnamed: 0  Id  SalePrice  MSSubClass  MSZoning  LotFrontage   LotArea  \
0           0   1  12.247694    0.235294      0.75     0.418208  0.366344   
1           1   2  12.109011    0.000000      0.75     0.495064  0.391317   
2           2   3  12.317167    0.235294      0.75     0.434909  0.422359   
3           3   4  11.849398    0.294118      0.75     0.388581  0.390295   
4           4   5  12.429216    0.235294      0.75     0.513123  0.468761   

   Street  Alley  LotShape  LandContour  Utilities  LotConfig  LandSlope  \
0     1.0    0.5  0.000000     0.333333        1.0       0.00        0.0   
1     1.0    0.5  0.000000     0.333333        1.0       0.50        0.0   
2     1.0    0.5  0.333333     0.333333        1.0       0.00        0.0   
3     1.0    0.5  0.333333     0.333333        1.0       0.25        0.0   
4     1.0    0.5  0.333333     0.333333        1.0       0.50        0.0   

   Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  \
0      0.681818         0.4         1.0      0.75         1.0     0.666667   
1      0.454545         0.2         1.0      0.75         0.6     0.555556   
2      0.681818         0.4         1.0      0.75         1.0     0.666667   
3      0.772727         0.4         1.0      0.75         1.0     0.666667   
4      0.954545         0.4         1.0      0.75         1.0     0.777778   

   OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  \
0        0.500   0.036765      0.098361        0.0       0.0          0.9   
1        0.875   0.227941      0.524590        0.0       0.0          0.4   
2        0.500   0.051471      0.114754        0.0       0.0          0.9   
3        0.500   0.669118      0.606557        0.0       0.0          0.3   
4        0.500   0.058824      0.147541        0.0       0.0          0.9   

   Exterior2nd  MasVnrType  MasVnrArea  ExterQual  ExterCond  Foundation  \
0          0.9        0.50     0.12250   0.666667        1.0        1.00   
1          0.3        0.25     0.00000   0.333333        1.0        0.50   
2          0.9        0.50     0.10125   0.666667        1.0        1.00   
3          0.2        0.25     0.00000   0.333333        1.0        0.25   
4          0.9        0.50     0.21875   0.666667        1.0        1.00   

   BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  BsmtFinSF1  BsmtFinType2  \
0      0.75      0.75          0.25      1.000000    0.125089      0.666667   
1      0.75      0.75          1.00      0.666667    0.173281      0.666667   
2      0.75      0.75          0.50      1.000000    0.086109      0.666667   
3      0.50      1.00          0.25      0.666667    0.038271      0.666667   
4      0.75      0.75          0.75      1.000000    0.116052      0.666667   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  Heating  HeatingQC  CentralAir  \
0         0.0   0.064212     0.140098      1.0       1.00         1.0   
1         0.0   0.121575     0.206547      1.0       1.00         1.0   
2         0.0   0.185788     0.150573      1.0       1.00         1.0   
3         0.0   0.231164     0.123732      1.0       0.75         1.0   
4         0.0   0.209760     0.187398      1.0       1.00         1.0   

   Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0         1.0  0.356155  0.413559           0.0   0.577712      0.333333   
1         1.0  0.503056  0.000000           0.0   0.470245      0.000000   
2         1.0  0.383441  0.419370           0.0   0.593095      0.333333   
3         1.0  0.399941  0.366102           0.0   0.579157      0.333333   
4         1.0  0.466237  0.509927           0.0   0.666523      0.333333   

   BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  \
0           0.0  0.666667       0.5         0.375      0.333333     0.666667   
1           0.5  0.666667       0.0         0.375      0.333333     0.333333   
2           0.0  0.666667       0.5         0.375      0.333333     0.666667   
3  

In [4]:
dataset.columns

Index(['Unnamed: 0', 'Id', 'SalePrice', 'MSSubClass', 'MSZoning',
       'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour',
       'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1',
       'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond',
       'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
       'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
       'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
     

In [5]:
train_Y=dataset[['SalePrice']]

In [6]:
train_X=dataset.drop(['Unnamed: 0','Id','SalePrice'],axis=1)

In [7]:
feature_sel_model=SelectFromModel(Lasso(alpha=0.005,random_state=0))
feature_sel_model.fit(train_X,train_Y)

SelectFromModel(estimator=Lasso(alpha=0.005, random_state=0))

In [8]:
feature_sel_model.get_support()

array([ True,  True, False, False, False, False, False, False, False,
       False, False,  True, False, False, False, False,  True, False,
       False,  True, False, False, False, False, False, False, False,
       False, False,  True, False,  True, False, False, False, False,
       False, False, False,  True,  True, False,  True, False, False,
        True,  True, False, False, False, False, False,  True, False,
       False,  True,  True,  True, False,  True,  True, False, False,
       False,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False])

In [9]:
# lets print the number of total and selected features
selected_feat=train_X.columns[(feature_sel_model.get_support())]

print('total features: {}'.format((train_X.shape[1])))
print('selected features: {}'.format(len(selected_feat)))

total features: 82
selected features: 19


In [10]:
selected_feat

Index(['MSSubClass', 'MSZoning', 'Neighborhood', 'OverallQual', 'YearRemodAdd',
       'BsmtQual', 'BsmtExposure', 'HeatingQC', 'CentralAir', '1stFlrSF',
       'GrLivArea', 'BsmtFullBath', 'KitchenQual', 'Fireplaces', 'FireplaceQu',
       'GarageType', 'GarageFinish', 'GarageCars', 'PavedDrive'],
      dtype='object')

In [11]:
train_X=train_X[selected_feat]

In [12]:
train_X.head()

MSSubClass  MSZoning  Neighborhood  OverallQual  YearRemodAdd  BsmtQual  \
0    0.235294      0.75      0.681818     0.666667      0.098361      0.75   
1    0.000000      0.75      0.454545     0.555556      0.524590      0.75   
2    0.235294      0.75      0.681818     0.666667      0.114754      0.75   
3    0.294118      0.75      0.772727     0.666667      0.606557      0.50   
4    0.235294      0.75      0.954545     0.777778      0.147541      0.75   

   BsmtExposure  HeatingQC  CentralAir  1stFlrSF  GrLivArea  BsmtFullBath  \
0          0.25       1.00         1.0  0.356155   0.577712      0.333333   
1          1.00       1.00         1.0  0.503056   0.470245      0.000000   
2          0.50       1.00         1.0  0.383441   0.593095      0.333333   
3          0.25       0.75         1.0  0.399941   0.579157      0.333333   
4          0.75       1.00         1.0  0.466237   0.666523      0.333333   

   KitchenQual  Fireplaces  FireplaceQu  GarageType  GarageFinish  GarageCars  \
0     0.666667    0.000000          0.2         0.8      0.666667        0.50   
1     0.333333    0.333333          0.6         0.8      0.666667        0.50   
2     0.666667    0.333333          0.6         0.8      0.666667        0.50   
3     0.666667    0.333333          0.8         0.4      0.333333        0.75   
4     0.666667    0.333333          0.6         0.8      0.666667        0.75   

   PavedDrive  
0         1.0  
1         1.0  
2         1.0  
3         1.0  
4         1.0

In [13]:
train_X.shape

(1460, 19)

In [14]:
data=pd.read_csv('test_X.csv')

In [15]:
data.head()

Id  MSSubClass  MSZoning  LotFrontage   LotArea  Street  Alley  LotShape  \
0  1461    0.000000       0.6     0.593445  0.566360     1.0    0.5       1.0   
1  1462    0.000000       0.8     0.598957  0.622527     1.0    0.5       0.0   
2  1463    0.235294       0.8     0.558854  0.614005     1.0    0.5       0.0   
3  1464    0.235294       0.8     0.582212  0.524583     1.0    0.5       0.0   
4  1465    0.588235       0.8     0.317987  0.335596     1.0    0.5       0.0   

   LandContour  Utilities  LotConfig  LandSlope  Neighborhood  Condition1  \
0     1.000000        0.0        1.0        0.0      0.500000       0.125   
1     1.000000        0.0        0.0        0.0      0.500000       0.250   
2     1.000000        0.0        1.0        0.0      0.333333       0.250   
3     1.000000        0.0        1.0        0.0      0.333333       0.250   
4     0.333333        0.0        1.0        0.0      0.916667       0.250   

   Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  \
0         0.5       0.0    0.333333     0.444444        0.625   0.384615   
1         0.5       0.0    0.333333     0.555556        0.625   0.407692   
2         0.5       0.0    0.666667     0.444444        0.500   0.107692   
3         0.5       0.0    0.666667     0.555556        0.625   0.100000   
4         0.5       1.0    0.333333     0.777778        0.500   0.146154   

   YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  Exterior2nd  MasVnrType  \
0      0.822581        0.2       0.0     0.846154     0.866667        0.75   
1      0.870968        0.6       0.0     0.923077     0.933333        0.25   
2      0.225806        0.2       0.0     0.846154     0.866667        0.75   
3      0.225806        0.2       0.0     0.846154     0.866667        0.25   
4      0.322581        0.2       0.0     0.461538     0.400000        0.75   

   MasVnrArea  ExterQual  ExterCond  Foundation  BsmtQual  BsmtCond  \
0    0.000000   1.000000        1.0         0.2       1.0       1.0   
1    0.083721   1.000000        1.0         0.2       1.0       1.0   
2    0.000000   1.000000        1.0         0.4       0.5       1.0   
3    0.015504   1.000000        1.0         0.4       1.0       1.0   
4    0.000000   0.666667        1.0         0.4       0.5       1.0   

   BsmtExposure  BsmtFinType1  BsmtFinSF1  BsmtFinType2  BsmtFinSF2  \
0           1.0      0.833333    0.116708           0.5    0.094364   
1           1.0      0.000000    0.230175           1.0    0.000000   
2           1.0      0.333333    0.197257           1.0    0.000000   
3           1.0      0.333333    0.150125           1.0    0.000000   
4           1.0      0.000000    0.065586           1.0    0.000000   

   BsmtUnfSF  TotalBsmtSF  Heating  HeatingQC  CentralAir  Electrical  \
0   0.126168     0.173111      0.0        1.0         1.0         1.0   
1   0.189720     0.260844      0.0        1.0         1.0         1.0   
2   0.064019     0.182139      0.0        0.5         1.0         1.0   
3   0.151402     0.181747      0.0        0.0         1.0         1.0   
4   0.475234     0.251227      0.0        0.0         1.0         1.0   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
0  0.312253  0.000000           0.0   0.312253           0.0           0.0   
1  0.468253  0.000000           0.0   0.468253           0.0           0.0   
2  0.326139  0.376477           0.0   0.548792           0.0           0.0   
3  0.325285  0.364125           0.0   0.542672           0.0           0.0   
4  0.453388  0.000000           0.0   0.453388           0.0           0.0   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  \
0      0.25       0.0      0.333333           0.5          1.0      0.166667   
1      0.25       0.5      0.500000           0.5          0.5      0.250000   
2      0.50       0.5      0.500000           0.5          1.0      0.250000   
3      0.50       0.5      0.500000           0

In [16]:
test_X=data[['MSSubClass', 'MSZoning', 'Neighborhood', 'OverallQual', 'YearRemodAdd',
       'BsmtQual', 'BsmtExposure', 'HeatingQC', 'CentralAir', '1stFlrSF',
       'GrLivArea', 'BsmtFullBath', 'KitchenQual', 'Fireplaces', 'FireplaceQu',
       'GarageType', 'GarageFinish', 'GarageCars', 'PavedDrive']]

In [17]:
test_X.head()

MSSubClass  MSZoning  Neighborhood  OverallQual  YearRemodAdd  BsmtQual  \
0    0.000000       0.6      0.500000     0.444444      0.822581       1.0   
1    0.000000       0.8      0.500000     0.555556      0.870968       1.0   
2    0.235294       0.8      0.333333     0.444444      0.225806       0.5   
3    0.235294       0.8      0.333333     0.555556      0.225806       1.0   
4    0.588235       0.8      0.916667     0.777778      0.322581       0.5   

   BsmtExposure  HeatingQC  CentralAir  1stFlrSF  GrLivArea  BsmtFullBath  \
0           1.0        1.0         1.0  0.312253   0.312253           0.0   
1           1.0        1.0         1.0  0.468253   0.468253           0.0   
2           1.0        0.5         1.0  0.326139   0.548792           0.0   
3           1.0        0.0         1.0  0.325285   0.542672           0.0   
4           1.0        0.0         1.0  0.453388   0.453388           0.0   

   KitchenQual  Fireplaces  FireplaceQu  GarageType  GarageFinish  GarageCars  \
0          1.0        0.00          0.6    0.166667      1.000000         0.2   
1          0.5        0.00          0.6    0.166667      1.000000         0.2   
2          1.0        0.25          1.0    0.166667      0.000000         0.4   
3          0.5        0.25          0.4    0.166667      0.000000         0.4   
4          0.5        0.00          0.6    0.166667      0.666667         0.4   

   PavedDrive  
0         1.0  
1         1.0  
2         1.0  
3         1.0  
4         1.0

In [18]:
test_X.shape

(1459, 19)

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor

In [20]:
X_train1, X_test1, y_train1,y_test1=train_test_split(train_X,train_Y)


In [21]:
print(X_train1.shape, X_test1.shape, y_train1.shape,y_test1.shape)

(1095, 19) (365, 19) (1095, 1) (365, 1)


In [22]:
y_train1.head()

SalePrice
312   11.694413
373   11.719940
1097  12.043554
544   12.098849
271   12.394625

In [23]:
y_test1.head()

SalePrice
420   12.237087
1123  11.678440
181   12.206573
610   12.653958
1296  11.951180

In [24]:
from sklearn.linear_model import LinearRegression

In [25]:
lr=LinearRegression()
lr.fit(X_train1,y_train1)

LinearRegression()

In [26]:
lr.score(X_train1,y_train1)

0.8911585541392028

In [27]:
pred=lr.predict(X_test1)

In [28]:
pred

array([[12.28676458],
       [11.44260887],
       [12.02122141],
       [12.74234352],
       [11.8609059 ],
       [11.63485241],
       [11.72191367],
       [11.73878986],
       [12.65302634],
       [11.9779882 ],
       [11.75580563],
       [11.93557465],
       [11.64856329],
       [12.49532796],
       [12.0596006 ],
       [11.71718999],
       [12.07398798],
       [11.98402248],
       [12.4209453 ],
       [12.45464894],
       [12.64565938],
       [12.01024766],
       [12.3831003 ],
       [12.19345622],
       [12.29332112],
       [11.6625533 ],
       [12.18997667],
       [11.83431396],
       [12.72707617],
       [12.45611547],
       [11.86646823],
       [12.2419693 ],
       [11.82163899],
       [12.52911359],
       [11.91350052],
       [11.87669889],
       [12.03760968],
       [12.21098499],
       [12.51563105],
       [12.10267417],
       [12.66023422],
       [12.45257148],
       [12.20535915],
       [12.2859152 ],
       [11.97467514],
       [11

In [29]:
lr.fit(train_X,train_Y)
lr.score(train_X,train_Y)

0.8835194845630763

In [30]:
pred=lr.predict(test_X)

In [31]:
pred

array([[11.77624332],
       [12.00142609],
       [12.01348255],
       ...,
       [12.094994  ],
       [11.64487645],
       [12.1676772 ]])

In [32]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge,Lasso
from sklearn.model_selection import cross_val_score

In [33]:
# Function to get cross validation scores
def get_cv_scores(model):
    scores=cross_val_score(model,train_X,train_Y,cv=5,scoring='r2')
    print('CV mean:',np.mean(scores))
    print('STD:',np.std(scores))
    print('\n')


In [34]:
lr=LinearRegression().fit(train_X,train_Y)
get_cv_scores(lr)

CV mean: 0.8781056878162502
STD: 0.016748521169936196




Regularization can be defined as explicitly restricting a model to prevent overfitting.
Lets apply regularization to check whether the model is overfitted or not.

In [35]:
alpha=1
ridge=Ridge(alpha=1).fit(train_X,train_Y)
get_cv_scores(ridge)

CV mean: 0.8777957190221567
STD: 0.015523212651199964




In [36]:
# find optimal alpha with grid search
alpha=[0.001,0.01,0.1,1,10,100,1000]
param_grid=dict(alpha=alpha)
grid=GridSearchCV(estimator=ridge,param_grid=param_grid,scoring='r2',verbose=1,n_jobs=-1)
grid_result=grid.fit(train_X,train_Y)
print('Best Score:', grid_result.best_score_)
print('Best params:',grid_result.best_params_ )

Fitting 5 folds for each of 7 candidates, totalling 35 fits
Best Score: 0.8781315694954998
Best params: {'alpha': 0.1}


In [37]:
alpha=1
lasso=Lasso(alpha=1).fit(train_X,train_Y)
get_cv_scores(lasso)

CV mean: -0.0034683244982841367
STD: 0.005131815546087139




In [38]:
# find optimal alpha with grid search
alpha=[0.001,0.01,0.1,1,10,100,1000]
param_grid=dict(alpha=alpha)
grid=GridSearchCV(estimator=lasso,param_grid=param_grid,scoring='r2',verbose=1,n_jobs=-1)
grid_result=grid.fit(train_X,train_Y)
print('Best Score:', grid_result.best_score_)
print('Best params:',grid_result.best_params_ )

Fitting 5 folds for each of 7 candidates, totalling 35 fits
Best Score: 0.8771871847384366
Best params: {'alpha': 0.001}


our score improved by optimizing alpha

Initially We got a r2 value of 0.8781 and standard deviation of 0.01674. The high value of r2 indicates our model is accurate and also it is clear that after applying ridge regularization our standard deviation little bit decreased which suggests we are less likely to be overfitting.